architecture/2025-03-upgrade-feature-spec.md

```diff
@@ -0,0 +1,226 @@
+
+- `helm install radius oci://ghcr.io/radius-project/helm-chart/radius --version 0.43.0`
+- `helm upgrade radius oci://ghcr.io/radius-project/helm-chart/radius --version 0.44.0`
+
+Users will still need to install Dapr and Contour and run `rad init` to achieve a similar behavior to what `rad upgrade kubernetes` would provide. I also think that the Radius CLI version needs to be updated to the desired version before running `rad init`.
+
+## Key dependencies and risks
+
+- **Dependency Name**: Helm – The upgrade process relies on Helm for managing the upgrade process. Issues/concerns/risks: Compatibility with new versions of Helm.
+- **Dependency Name**: Contour – The upgrade process relies on Contour for routing requests. Issues/concerns/risks: Compatibility with new versions of Contour.
+- **Dependency Name**: Dapr – The upgrade process relies on Dapr. Issues/concerns/risks: Compatibility with new versions of Dapr.
+
+## Key assumptions to test and questions to answer
+
+### Core Assumptions
+
+- **Version Compatibility:** If you are upgrading from a version of Radius that does not include the `rad upgrade kubernetes` command, you must first update to a release that provides the upgrade functionality. Older versions do not support automated upgrades in the CLI.
+- **User Permissions:** Users have the necessary permissions and access to perform upgrades, including cluster-admin roles where required.
+- **Data Integrity:** The upgrade process will maintain data integrity and prevent corruption during migration operations (particularly for database migrations) by using snapshots and the ability to restore in case of failure.
+- **Dependency Management:** We use a specific version of Contour and Dapr (<https://github.com/radius-project/radius/blob/main/pkg/cli/helm/cluster.go#L34>) and they can not be updated with the configuration. This ensures consistent dependency management. There is a case where the default chart versions may be updated in a newer version of Radius but that means that those versions of Contour and Dapr are probably tested with that new version of Radius and shouldn't provide problems during the upgrade from the current version and the desired version which is the newer version that we just discussed about.
+- **Resource Requirements:** The upgrade process won't exceed available cluster resources during the transition period when both old and new components may be running simultaneously.
+
+### Technical Questions to Resolve
+
+- **Downgrade Support:**
+  - Should we support downgrading to previous versions? If yes, what are the limitations?
```

Comment:
I think we should support them to the same extent of the mechanisms used to support upgrades (e.g. preflight checks, rollback if something goes wrong, etc.)

Reply:
These questions and their answers are being discussed in the design document.

Reply:
I removed this section.